libcontainer/configs/validate/validator.go

```diff
@@ -273,6 +273,26 @@ func cgroupsCheck(config *configs.Config) error {
 	return nil
 }
 
+func checkBindOptions(m *configs.Mount) error {
+	if m.Flags&unix.MS_BIND == 0 {
+		return nil
+	}
+	// We must reject bind-mounts that also have filesystem-specific mount
+	// options, because the kernel will completely ignore these flags and we
+	// cannot set them per-mountpoint.
+	//
+	// It should be noted that (due to how the kernel caches superblocks), data
+	// options could also silently ignored for other filesystems even when
+	// doing a fresh mount, but there is no real way to avoid this (and it
+	// matches how everything else works). There have been proposals to make it
+	// possible for userspace to detect this caching, but this wouldn't help
+	// runc because the behaviour wouldn't even be desirable for most users.
+	if m.Data != "" {
+		return fmt.Errorf("bind mounts cannot have any filesystem-specific options applied")
```

Comment:
What does filesystem-specific mean here? In the runtime spec I don't see any differentiation between fs-specific and not specific flags (https://github.com/opencontainers/runtime-spec/blob/main/config.md#linux-mount-options). Am I missing something?

It will be nice to throw a clearer error for the user here. **The` idmap` and `ridmap` flags have just been added to that table and those will be valid, while the rest not**. It will be nice if we can throw a better error or separate the flags in the spec, so a user can know what is a valid combination and what it is not.

Reply:
In fact, runtime spec mentions both "filesystem-specific" and "filesystem-independent" mount options, referring to `mount(8)` (and the only "filesystemd-specific" options it mentions are mount propagation modes for bind mounts).

I guess we need to change the wording in the spec to distinguish the real filesystem-specific mount options (not documented in runtime-spec at all) from the bind mount propagation options. WDYT @AkihiroSuda ?

Reply:
> I guess we need to change the wording in the spec to distinguish the real filesystem-specific mount options (not documented in runtime-spec at all) from the bind mount propagation options. WDYT @AkihiroSuda ?

SGTM, PR is wanted

Reply:
> I guess we need to change the wording in the spec to distinguish the real filesystem-specific mount options (not documented in runtime-spec at all) from the bind mount propagation options. 

Opened https://github.com/opencontainers/runtime-spec/pull/1228

Reply:
@rata as for this PR, I think the error message wording ("bind mounts cannot have any filesystem-specific options applied") is totally correct.

Reply:
@kolyshkin thanks for the PR! SGTM this error, then  :)